pkg/applicationserver/cleanup.go

```diff
@@ -0,0 +1,67 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package applicationserver
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/cleanup"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/unique"
+)
+
+type RegistryCleaner struct {
+	DevRegistry    DeviceRegistry
+	AppUpsRegistry ApplicationUplinkRegistry
+	LocalSet       map[string]struct{}
+}
+
+func (cleaner *RegistryCleaner) RangeToLocalSet(ctx context.Context) error {
+	cleaner.LocalSet = make(map[string]struct{})
+	err := cleaner.DevRegistry.Range(ctx, []string{"ids"}, func(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, dev *ttnpb.EndDevice) bool {
+		cleaner.LocalSet[unique.ID(ctx, ids)] = struct{}{}
+		return true
+	},
+	)
+	return err
+}
+
+func (cleaner *RegistryCleaner) DeleteComplement(ctx context.Context, devSet map[string]struct{}) error {
+	for ids := range devSet {
+		devIds, err := unique.ToDeviceID(ids)
+		if err != nil {
+			return err
+		}
```

Comment:
Use `unique.WithContext` afterwards.
```
	ctx := ctx
	ctx, err = unique.WithContext(ctx, uid)
	if err != nil {
		return nil, err
	}
```
In order to fill the context with the unique data. This should be done after each `unique.ToDeviceID`.